File: benchmarks/as/assembly/concepts/inClose.ts
import { parseBurmeister } from "../parsing/burmeister";
import { createQueue, dequeue, enqueue, isQueueEmpty } from "../structures/queue";
import { FormalConcept } from "../types/FormalConcept";
import { FormalContext, formalContextHasAttribute, hasObjectWithAllAttributes } from "../types/FormalContext";
import { LinkedList } from "../types/LinkedList";
import { FormalConceptsTimedResult } from "../types/TimedResult";
import { copyArray, copyStaticArray } from "../utils/arrays";

export function inCloseBurmeister(burmeisterContext: string): FormalConceptsTimedResult {
    const context = parseBurmeister(burmeisterContext);
    return inClose(context);
}

export function inClose(context: FormalContext): FormalConceptsTimedResult {
    const startTime = Date.now();

    const initialConceptObjects = new StaticArray<i32>(context.objects.length);
    for (let i = 0; i < context.objects.length; i++)
        initialConceptObjects[i] = i;

    const newExtentBuffer = new StaticArray<i32>(context.objects.length);
    const initialConcept: FormalConcept = { attributes: [], objects: initialConceptObjects, attribute: 0 };
    const formalConcepts = new Array<FormalConcept>();
    formalConcepts.push(initialConcept);

    inCloseImpl(
        context,
        newExtentBuffer,
        formalConcepts,
        initialConcept,
        0);

    if (!hasObjectWithAllAttributes(context)) {
        const conceptAttributes = new Array<i32>(context.attributes.length);
        for (let i = 0; i < context.attributes.length; i++)
            conceptAttributes[i] = i;

        formalConcepts.push({ attributes: conceptAttributes, objects: [], attribute: 0 });
    }

    return {
        value: formalConcepts,
        time: Date.now() - startTime,
    };
}

function inCloseImpl(
    context: FormalContext,
    newExtentBuffer: StaticArray<i32>,
    formalConcepts: Array<FormalConcept>,
    parentConcept: FormalConcept,
    currentAttribute: i32
): void {
    const queue: LinkedList<FormalConcept> = createQueue<FormalConcept>();

    for (let j = currentAttribute; j < context.attributes.length; j++) {
        let lastObjectIndex: i32 = 0;

        // Take those objects from the parentConcept that have attribute j, i.e. generate a new potential extent
        for (let i = 0; i < parentConcept.objects.length; i++) {
            const object = parentConcept.objects[i];
            if (formalContextHasAttribute(context, object, j)) {
                newExtentBuffer[lastObjectIndex] = object;
                lastObjectIndex++;
            }
        }

        if (lastObjectIndex > 0) {
            if (lastObjectIndex === parentConcept.objects.length) {
                parentConcept.attributes.push(j);
            }
            else if (isCannonical(context, parentConcept, newExtentBuffer, lastObjectIndex, j - 1)) {
                const newIntent = copyArray<i32>(parentConcept.attributes, parentConcept.attributes.length + 1);
                newIntent[newIntent.length - 1] = j;
                const newConcept: FormalConcept = {
                    attributes: newIntent,
                    objects: copyStaticArray<i32>(newExtentBuffer, lastObjectIndex),
                    attribute: j
                };
                formalConcepts.push(newConcept);

                enqueue<FormalConcept>(queue, newConcept);
                //inCloseImpl(context, newExtentBuffer, formalConcepts, newConcept, j + 1);
            }
        }
    }

    while (!isQueueEmpty(queue)) {
        const concept = dequeue<FormalConcept>(queue, parentConcept);

        if (concept === parentConcept) {
            return;
        }

        inCloseImpl(context, newExtentBuffer, formalConcepts, concept, concept.attribute + 1);
    }
}

function isCannonical(
    context: FormalContext,
    parentConcept: FormalConcept,
    newExtentBuffer: StaticArray<i32>,
    newExtentSize: i32,
    startingAttribute: i32,
): boolean {
    for (let k = parentConcept.attributes.length - 1; k >= 0; k--) {
        for (let j = startingAttribute; j >= parentConcept.attributes[k] + 1; j--) {
            let h: i32;

            for (h = 0; h < newExtentSize; h++) {
                if (!formalContextHasAttribute(context, newExtentBuffer[h], j))
                    break;
            }
            if (h === newExtentSize) {
                return false;
            }
        }
        startingAttribute = parentConcept.attributes[k] - 1;
    }

    for (let j = startingAttribute; j >= 0; j--) {
        let h: i32;

        for (h = 0; h < newExtentSize; h++) {
            if (!formalContextHasAttribute(context, newExtentBuffer[h], j))
                break;
        }
        if (h === newExtentSize) {
            return false;
        }
    }

    return true;
}

/*
export function inClose(context: FormalContext): Array<FormalConcept> {
    const startTime = Date.now();

    const initialConceptObjects = new StaticArray<i32>(context.objects.length);
    for (let i = 0; i < context.objects.length; i++)
        initialConceptObjects[i] = i;

    const newExtentBuffer = new StaticArray<i32>(context.objects.length);
    const initialConcept: FormalConcept = { attributes: [], objects: initialConceptObjects, attribute: 0 };
    const formalConcepts = new Array<FormalConcept>();
    formalConcepts.push(initialConcept);
    
    inCloseImpl(
        context,
        newExtentBuffer,
        formalConcepts,
        initialConcept,
        0,
        createZeroIntent(context.attributes.length));

    if (!hasObjectWithAllAttributes(context)) {
        const conceptAttributes = new Array<i32>(context.attributes.length);
        for (let i = 0; i < context.attributes.length; i++)
            conceptAttributes[i] = i;

        formalConcepts.push({ attributes: conceptAttributes, objects: [], attribute: 0 });
    }

    console.log(`InClose: ${Date.now() - startTime} ms`);

    return formalConcepts;
}

function inCloseImpl(
    context: FormalContext,
    newExtentBuffer: StaticArray<i32>,
    formalConcepts: Array<FormalConcept>,
    parentConcept: FormalConcept,
    currentAttribute: i32,
    booleanParentIntent: StaticArray<u64>
): void {
    const queue: LinkedList<FormalConcept> = createQueue<FormalConcept>();

    for (let j = currentAttribute; j < context.attributes.length; j++) {
        if (!intentHasAttribute(booleanParentIntent, j)) {
            let lastObjectIndex: i32 = 0;

            // Take those objects from the parentConcept that have attribute j, i.e. generate a new potential extent
            for (let i = 0; i < parentConcept.objects.length; i++) {
                const object = parentConcept.objects[i];
                if (formalContextHasAttribute(context, object, j)) {
                    newExtentBuffer[lastObjectIndex] = object;
                    lastObjectIndex++;
                }
            }

            if (lastObjectIndex === 0) {
                addAttributeToIntent(booleanParentIntent, j);
            }
            else {
                if (lastObjectIndex < parentConcept.objects.length) {
                    if (isCannonical(context, parentConcept, newExtentBuffer, lastObjectIndex, j - 1)) {
                        const newIntent = copyArray<i32>(parentConcept.attributes, parentConcept.attributes.length + 1);
                        newIntent[newIntent.length - 1] = j;
                        const newConcept: FormalConcept = {
                            attributes: newIntent,
                            objects: copyStaticArray<i32>(newExtentBuffer, lastObjectIndex),
                            attribute: j
                        };
                        formalConcepts.push(newConcept);
    
                        enqueue<FormalConcept>(queue, newConcept);
                        //inCloseImpl(context, newExtentBuffer, formalConcepts, newConcept, j + 1);
                    }
                }
                else {
                    parentConcept.attributes.push(j);
                    addAttributeToIntent(booleanParentIntent, j);
                }
            }
        }
    }

    while (!isQueueEmpty(queue)) {
        const concept = dequeue<FormalConcept>(queue, parentConcept);

        if (concept === parentConcept) {
            return;
        }

        const booleanChildIntent = copyStaticArray<u64>(booleanParentIntent, booleanParentIntent.length);
        addAttributeToIntent(booleanChildIntent, concept.attribute);

        inCloseImpl(context, newExtentBuffer, formalConcepts, concept, concept.attribute + 1, booleanChildIntent);
    }
}

function isCannonical(
    context: FormalContext,
    parentConcept: FormalConcept,
    newExtentBuffer: StaticArray<i32>,
    newExtentSize: i32,
    startingAttribute: i32,
): boolean {
    for (let k = parentConcept.attributes.length - 1; k >= 0; k--) {
        for (let j = startingAttribute; j >= parentConcept.attributes[k] + 1; j--) {
            let h: i32;

            for (h = 0; h < newExtentSize; h++) {
                if (!formalContextHasAttribute(context, newExtentBuffer[h], j))
                    break;
            }
            if (h === newExtentSize) {
                return false;
            }
        }
        startingAttribute = parentConcept.attributes[k] - 1;
    }

    for (let j = startingAttribute; j >= 0; j--) {
        let h: i32;

        for (h = 0; h < newExtentSize; h++) {
            if (!formalContextHasAttribute(context, newExtentBuffer[h], j))
                break;
        }
        if (h === newExtentSize) {
            return false;
        }
    }

    return true;
}

function createZeroIntent(attributesCount: i32): StaticArray<u64> {
    const cellsCount = <i32>Math.ceil(<f64>attributesCount / BOOLEAN_CELL_SIZE);
    const intent = new StaticArray<u64>(cellsCount);

    for (let i = 0; i < intent.length; i++) {
        intent[i] = 0;
    }

    return intent;
}

function intentHasAttribute(intent: StaticArray<u64>, attribute: i32): boolean {
    const cell = attribute / BOOLEAN_CELL_SIZE;
    const cellValue: u64 = intent[cell];
    const mask: u64 = 1 << (attribute % BOOLEAN_CELL_SIZE);

    return (cellValue & mask) !== 0;
}

function addAttributeToIntent(intent: StaticArray<u64>, attribute: i32): void {
    const cell = attribute / BOOLEAN_CELL_SIZE;
    const cellValue: u64 = intent[cell];
    const mask: u64 = 1 << (attribute % BOOLEAN_CELL_SIZE);

    intent[cell] = cellValue | mask;
}
*/

File: src/cpp/inClose.cpp
// Implementation of the InClose algorithm:
// - https://www.researchgate.net/publication/228522038_In-Close_a_fast_algorithm_for_computing_formal_concepts

#include "utils.h"
#include "inClose.h"
#include "types/FormalConcept.h"
#include "types/TimedResult.h"

#include <iostream>
#include <memory>
#include <queue>
#include <vector>


// It is annoying to pass the formal context data to the functions like this,
// but it is the fastest and simplest way

/**
 * Checks if a concept has already been generated.
 * Returns `false` (not canonical) if the extent is found elsewhere.
 * Basically one-to-one implementation of the 3.2 section from the paper.
 */
bool isCannonical(
    std::vector<unsigned int>& contextMatrix,
    int cellSize,
    int cellsPerObject,
    FormalConcept& parentConcept,
    std::vector<int>& newExtentBuffer,
    int newExtentSize,
    int startingAttribute
) {
    std::vector<int>& parentConceptAttributes = parentConcept.getAttributes();

    // Iterate through blocks of columns between those already in the intent
    for (int k = parentConceptAttributes.size() - 1; k >= 0; k--) {
        for (int j = startingAttribute; j >= parentConceptAttributes[k] + 1; j--) {
            int h = 0;

            // Check if the current attribute j is shared by all objects in the new extent
            for (h = 0; h < newExtentSize; h++) {
                if (!formalContextHasAttribute(
                    contextMatrix,
                    cellSize,
                    cellsPerObject,
                    newExtentBuffer[h],
                    j
                ))
                    break;
            }
            // If the extent is found (all objects share attribute j), it's not canonical
            if (h == newExtentSize) {
                return false;
            }
        }
        // Prepare to skip the column just iterated down to
        startingAttribute = parentConceptAttributes[k] - 1;
    }

    // Final search for the extent in the block of columns down to 0
    for (int j = startingAttribute; j >= 0; j--) {
        int h = 0;

        for (h = 0; h < newExtentSize; h++) {
            if (!formalContextHasAttribute(
                contextMatrix,
                cellSize,
                cellsPerObject,
                newExtentBuffer[h],
                j
            ))
                break;
        }
        if (h == newExtentSize) {
            return false;
        }
    }

    return true;
}

/**
 * Core recursive implementation of In-Close.
 * Implementation of the 3.1 section from the paper.
 */
void inCloseImpl(
    std::vector<unsigned int>& contextMatrix,
    int cellSize,
    int cellsPerObject,
    int contextObjectsCount,
    int contextAttributesCount,
    std::vector<int>& newExtentBuffer,
    std::vector<FormalConcept>& formalConcepts,
    int parentConceptIndex,
    int currentAttribute
#ifdef __EMSCRIPTEN__
    , OnProgressCallback& onProgress,
    bool callOnProgress
#endif
) {
    std::queue<int> conceptsQueue;

    // Iterate through attributes starting from the currentAttribute
    for (int j = currentAttribute; j < contextAttributesCount; j++) {
        int lastObjectIndex = 0;
        std::vector<int>& parentConceptObjects = formalConcepts[parentConceptIndex].getObjects();

        // Take those objects from the parentConcept that have attribute j, i.e. generate a new potential extent
        for (int i = 0; i < parentConceptObjects.size(); i++) {
            int object = parentConceptObjects[i];

            if (formalContextHasAttribute(contextMatrix, cellSize, cellsPerObject, object, j)) {
                newExtentBuffer[lastObjectIndex] = object;
                lastObjectIndex++;
            }
        }

        if (lastObjectIndex > 0) {
            // Case A: Extent didn't change -> Attribute j belongs to the current concept's intent
            if (lastObjectIndex == parentConceptObjects.size()) {
                std::vector<int>& attributes = formalConcepts[parentConceptIndex].getAttributes();
                attributes.push_back(j);
            }
            // Case B: Extent is a smaller non-empty intersection -> Check if this forms a new canonical concept
            else if (isCannonical(
                contextMatrix,
                cellSize,
                cellsPerObject,
                formalConcepts[parentConceptIndex],
                newExtentBuffer,
                lastObjectIndex,
                j - 1
            )) {
                formalConcepts.emplace_back();
                FormalConcept& newConcept = formalConcepts.back();

                // Inherit parent attributes and add the new qualifying attribute j
                std::vector<int> newIntent = formalConcepts[parentConceptIndex].getAttributesCopy();
                newIntent.push_back(j);
                newConcept.setAttributes(newIntent);

                // Set the newly filtered objects as the extent
                std::vector<int> newExtent(newExtentBuffer.begin(), newExtentBuffer.begin() + lastObjectIndex);
                newConcept.setObjects(newExtent);

                newConcept.setAttribute(j);

                // Queue this concept for further recursive calls
                conceptsQueue.push(formalConcepts.size() - 1);
            }
        }
    }

#ifdef __EMSCRIPTEN__
    int progressCounter = 0;
    int progressStepsCount = conceptsQueue.size() + 1;
#endif

    // Recursively process each new concept discovered in the loop above
    while (!conceptsQueue.empty()) {
        int conceptIndex = conceptsQueue.front();

        inCloseImpl(
            contextMatrix,
            cellSize,
            cellsPerObject,
            contextObjectsCount,
            contextAttributesCount,
            newExtentBuffer,
            formalConcepts,
            conceptIndex,
            formalConcepts[conceptIndex].getAttribute() + 1
#ifdef __EMSCRIPTEN__
            , onProgress,
            false
#endif
        );

#ifdef __EMSCRIPTEN__
        if (callOnProgress && !onProgress.isUndefined()) {
            progressCounter++;
            onProgress((double)progressCounter / progressStepsCount);
        }
#endif

        conceptsQueue.pop();
    }
}

void inClose(
    TimedResult<std::vector<FormalConcept>>& result,
    std::vector<unsigned int>& contextMatrix,
    int cellSize,
    int cellsPerObject,
    int contextObjectsCount,
    int contextAttributesCount
#ifdef __EMSCRIPTEN__
    , OnProgressCallback onProgress
#endif
) {
    //printFormalContext(contextMatrix, cellSize, cellsPerObject, contextObjectsCount, contextAttributesCount);

    long long startTime = nowMills();

    // Temporary buffer to reuse memory for extent calculations
    std::vector<int> newExtentBuffer;
    newExtentBuffer.resize(contextObjectsCount);

    // Create the initial concept: contains all objects, intent is empty
    std::vector<int> initialConceptObjects;
    initialConceptObjects.reserve(contextObjectsCount);
    for (int i = 0; i < contextObjectsCount; i++) {
        initialConceptObjects.push_back(i);
    }

    FormalConcept initialConcept = FormalConcept();
    initialConcept.setObjects(initialConceptObjects);
    initialConcept.setAttribute(0);

    result.value.push_back(initialConcept);

    // Start recursive processing from the first concept (index 0)
    inCloseImpl(
        contextMatrix,
        cellSize,
        cellsPerObject,
        contextObjectsCount,
        contextAttributesCount,
        newExtentBuffer,
        result.value,
        0,
        0
#ifdef __EMSCRIPTEN__
        , onProgress,
        true
#endif
        );

    // Edge case: Handle the concept containing all attributes if no object possesses all of them
    if (!hasObjectWithAllAttributes(
        contextMatrix,
        cellSize,
        cellsPerObject,
        contextObjectsCount,
        contextAttributesCount
    )) {
        std::vector<int> conceptAttributes;
        conceptAttributes.resize(contextAttributesCount);
        for (int i = 0; i < contextAttributesCount; i++) {
            conceptAttributes[i] = i;
        }

        FormalConcept allAttributesConcept = FormalConcept();
        allAttributesConcept.setAttributes(conceptAttributes);
        allAttributesConcept.setAttribute(0);

        result.value.push_back(allAttributesConcept);
    }

    long long endTime = nowMills();

#ifdef __EMSCRIPTEN__
    if (!onProgress.isUndefined()) {
        onProgress(1);
    }
#endif

    result.time = (int)endTime - startTime;
}
